### PR TITLE
Follow ups to DOPE-235, DOPE-226, DOPE-188, DOPE-228, DOPE-248 and DOPE-229 

### PR DESCRIPTION
**Pull request by martinagallati-ergon**
follow up PR to the last five PRs, containing the feedback from @pgruntz in #33, #34, #35, #36, #37, #38.

### CODE REVIEW

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArraySymmetricDifferenceExpression.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    vararg additionalArrays: TypeExpression<ArrayType<T>>,
+) : ArrayFunctionExpression<T>("ARRAY_SYMDIFF", firstArray, secondArray, *additionalArrays)
+
+fun <T : ValidType> arraySymDiff(
+    firstArray: TypeExpression<ArrayType<T>>,
+    secondArray: TypeExpression<ArrayType<T>>,
+    vararg additionalArrays: TypeExpression<ArrayType<T>>,
+) = ArraySymmetricDifferenceExpression(firstArray, secondArray, *additionalArrays)
+
+// only an alias (defined by couchbase)
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArraySymmetricDifferenceExpression.kt:19]: Well, the couchbase-function _ARRAY_SYMDIFF1_ is an alias of the couchabse-function _ARRAY_SYMDIFF_, but in dope apparently we never create the _ARRAY_SYMDIFF1_-String. We should add a ArraySymmetricDifference1Expression, if we want to support that function.
And that symdiff1 is an alias of something is also not relevant in the perspective of dope: We just allow the creation of the couchbase-functions and do not care about the actual implementation (just the parameters and the return value)

File: core/src/main/kotlin/ch/ergon/dope/resolvable/operator/FunctionOperator.kt [hunk 2]
```diff
@@ -4,14 +4,7 @@ import ch.ergon.dope.DopeQuery
...
 interface FunctionOperator {
-    fun toFunctionQueryString(symbol: String, vararg arguments: DopeQuery, extra: DopeQuery?) =
-        if (extra == null) {
-            toFunctionQueryString(symbol = symbol, *arguments)
-        } else {
-            toFunctionQueryString(symbol = symbol, *arguments, extra)
-        }
-
-    fun toFunctionQueryString(symbol: String, vararg arguments: DopeQuery) = arguments.joinToString(
+    fun toFunctionQueryString(symbol: String, vararg arguments: DopeQuery?) = arguments.filterNotNull().joinToString(
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/operator/FunctionOperator.kt:7]: small thing, can you remove in the function below:
`fun toFunctionQueryString(symbol: String, quantifier: AggregateQuantifier?, vararg arguments: String) =`
the _vararg_ from the last argument? I think it is always called with one argument

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayContainsExpression.kt [hunk 3]
```diff
@@ -1,15 +1,28 @@
...
+        val arrayDopeQuery = array.toDopeQuery()
+        val valueDopeQuery = value.toDopeQuery()
+        return DopeQuery(
+            queryString = toFunctionQueryString("ARRAY_CONTAINS", arrayDopeQuery, valueDopeQuery),
+            parameters = arrayDopeQuery.parameters + valueDopeQuery.parameters,
+        )
+    }
+}
 
 fun <T : ValidType> arrayContains(array: TypeExpression<ArrayType<T>>, value: TypeExpression<T>) =
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayContainsExpression.kt:27]: I think these functions should all be extensiosn-functions instead, something like:
`fun <T : ValidType> TypeExpression<ArrayType<T>>.contains(value: TypeExpression<T>) = ArrayContainsExpression(this, value)`
(same for all the other functions)
You should be able to write something like
```
.select(someStringArrayField().prepend("lol").min())
.where(someIntArrayField().concat(someIntArrayField()).contains(12))
```
But maybe you can do that in another PR

**Review by pgruntz — APPROVED** [review]

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayRangeExpression.kt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        val endDopeQuery = end.toDopeQuery()
+        val stepDopeQuery = step?.toDopeQuery()
+        return DopeQuery(
+            queryString = toFunctionQueryString("ARRAY_RANGE", startDopeQuery, endDopeQuery, stepDopeQuery),
+            parameters = startDopeQuery.parameters + endDopeQuery.parameters + stepDopeQuery?.parameters.orEmpty(),
+        )
+    }
+}
+
+fun arrayRange(
```
[the remark this answers is not in the document]
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayRangeExpression.kt:26]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/operator/FunctionOperator.kt (hunk 2 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/operator/FunctionOperator.kt:7]: Yes I can 😄 and I did

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayContainsExpression.kt (hunk 3 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayContainsExpression.kt:27]: I am going to do that in another PR, as to not inflate this PR further.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArraySymmetricDifferenceExpression.kt (hunk 1 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArraySymmetricDifferenceExpression.kt]: I changed the two instances of aliased functions, so now SYMDIFF1 is possible.

File: core/src/test/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayRangeExpressionTest.kt [hunk 5]
```diff
@@ -141,6 +177,19 @@ class ArrayRangeExpressionTest {
...
 
+    @Test
+    fun `should support ARRAY_RANGE extension number number type`() {
+        val start = someNumber()
+        val end = someNumber()
+        val step = someNumberField()
+        val expected = ArrayRangeExpression(start.toDopeType(), end.toDopeType(), step)
+
+        val actual = arrayRange(start, end, step)
+        // val new = arrayRange(start, end) goht nöd :(
```
Comment (pgruntz) [inline core/src/test/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayRangeExpressionTest.kt:188]: why not?

Comment (martinagallati-ergon) [reply core/src/test/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/arrayfunction/ArrayRangeExpressionTest.kt:188]: removed it 😄
